File: desktop/ui/src/audio-cycle-prewarm-coordinator.ts
export const AUDIO_CYCLE_COMMIT_GRACE_MS = 500;

export type AudioCycleCandidateStatus = 'planned' | 'preparing' | 'prepared' | 'committing';

export type AudioCycleCandidatePlan<T> = {
  candidateId: number;
  sample: T;
  targetAbsolutePositionMs: number;
  status: AudioCycleCandidateStatus;
};

export type AudioCycleCoordinatorAction = 'prepare' | 'commit' | 'expire' | null;

export function createAudioCycleCandidatePlan<T>(
  candidateId: number,
  sample: T,
  baseAbsolutePositionMs: number,
  periodMediaMs: number,
): AudioCycleCandidatePlan<T> {
  const safeBaseMs = Math.max(0, Math.round(baseAbsolutePositionMs));
  const safePeriodMs = Math.max(1, Math.round(periodMediaMs));
  return {
    candidateId,
    sample,
    targetAbsolutePositionMs: safeBaseMs + safePeriodMs,
    status: 'planned',
  };
}

export function getAudioCycleCoordinatorAction<T>(
  plan: AudioCycleCandidatePlan<T> | null,
  currentAbsolutePositionMs: number,
  playbackRate: number,
  queuedPlaybackMs = 0,
): AudioCycleCoordinatorAction {
  if (!plan) return null;
  const safePlaybackRate = Number.isFinite(playbackRate) && playbackRate > 0 ? playbackRate : 1;
  const safeQueuedPlaybackMs = Number.isFinite(queuedPlaybackMs) && queuedPlaybackMs > 0
    ? queuedPlaybackMs
    : 0;
  const graceMediaMs = Math.round(AUDIO_CYCLE_COMMIT_GRACE_MS * safePlaybackRate);
  if (plan.status === 'committing') return null;
  if (currentAbsolutePositionMs > plan.targetAbsolutePositionMs + graceMediaMs) return 'expire';
  const mediaLeadMs = plan.targetAbsolutePositionMs - currentAbsolutePositionMs;
  if (plan.status === 'planned') return 'prepare';
  const commitLeadMediaMs = Math.round(safeQueuedPlaybackMs * safePlaybackRate);
  if (plan.status === 'prepared' && mediaLeadMs <= commitLeadMediaMs) return 'commit';
  return null;
}

export function updateAudioCycleCandidateStatus<T>(
  plan: AudioCycleCandidatePlan<T>,
  status: AudioCycleCandidateStatus,
): AudioCycleCandidatePlan<T> {
  return { ...plan, status };
}

export type AudioCycleCommandMessage = {
  version: 1;
  type: 'audio-cycle-command';
  action: 'prepare' | 'commit' | 'cancel';
  candidate_id: number;
  playback_generation: number;
  base_audio_stream_revision?: number;
  target_absolute_position_ms?: number;
  audio?: Record<string, unknown>;
  audio_variants?: Record<string, unknown>[];
};

export type AudioCycleResultMessage = {
  version: 1;
  type: 'audio-cycle-result';
  action: 'prepare' | 'commit' | 'cancel';
  candidate_id: number;
  accepted: boolean;
  committed: boolean;
  reason: string | null;
  error_code?: string | null;
  snapshot?: unknown;
};

export function isAudioCycleCommandMessage(value: unknown): value is AudioCycleCommandMessage {
  if (!value || typeof value !== 'object') return false;
  const message = value as Partial<AudioCycleCommandMessage>;
  const validEnvelope = message.version === 1
    && message.type === 'audio-cycle-command'
    && ['prepare', 'commit', 'cancel'].includes(message.action ?? '')
    && Number.isSafeInteger(message.candidate_id)
    && Number(message.candidate_id) > 0
    && Number.isSafeInteger(message.playback_generation);
  if (!validEnvelope || message.action !== 'prepare') return validEnvelope;
  return Number.isSafeInteger(message.base_audio_stream_revision)
    && Number.isSafeInteger(message.target_absolute_position_ms)
    && Number(message.target_absolute_position_ms) >= 0
    && Boolean(message.audio)
    && typeof message.audio === 'object'
    && Array.isArray(message.audio_variants);
}

export function isAudioCycleResultMessage(value: unknown): value is AudioCycleResultMessage {
  if (!value || typeof value !== 'object') return false;
  const message = value as Partial<AudioCycleResultMessage>;
  return message.version === 1
    && message.type === 'audio-cycle-result'
    && ['prepare', 'commit', 'cancel'].includes(message.action ?? '')
    && Number.isSafeInteger(message.candidate_id)
    && typeof message.accepted === 'boolean'
    && typeof message.committed === 'boolean'
    && (message.reason === null || typeof message.reason === 'string')
    && (message.error_code == null || typeof message.error_code === 'string');
}

export function isExpectedAudioCycleSkipCode(code: string | null | undefined): boolean {
  return code === 'audio_mixer_candidate_silent' || code === 'audio_mixer_candidate_stale';
}

File: desktop-csharp-windows/src/GpAutoLive.Media/AudioCyclePrewarmCoordinator.cs
namespace GpAutoLive.Media;

/// <summary>N/N+1 音频候选的受限生命周期。</summary>
public enum AudioCycleCandidateStatus
{
    Planned,
    Preparing,
    Prepared,
    Committing,
}

/// <summary>当前候选调度器允许的单步动作。</summary>
public enum AudioCycleCoordinatorAction
{
    None,
    Prepare,
    Commit,
    Expire,
}

/// <summary>
/// 一个 N+1 候选计划。计划只保存调用方提供的候选引用和绝对媒体时间，
/// 不拥有解码任务、PCM 队列或无限期缓存。
/// </summary>
public sealed record AudioCycleCandidatePlan<T>(
    ulong CandidateId,
    T Sample,
    long TargetAbsolutePositionMs,
    AudioCycleCandidateStatus Status);

/// <summary>
/// 依据绝对媒体时间协调 N+1 的准备、提交和过期；N+2 不在此处创建计划。
/// </summary>
public static class AudioCyclePrewarmCoordinator
{
    public const long CommitGraceMediaMs = 500;

    /// <summary>创建计划时把基准和周期限制为有限的非负毫秒值。</summary>
    public static AudioCycleCandidatePlan<T> CreateCandidatePlan<T>(
        ulong candidateId,
        T sample,
        double baseAbsolutePositionMs,
        double periodMediaMs)
    {
        if (candidateId == 0)
        {
            throw new ArgumentOutOfRangeException(nameof(candidateId), "音频候选 ID 必须大于 0。");
        }

        var safeBase = NormalizeNonNegativeMilliseconds(baseAbsolutePositionMs, fallback: 0);
        var safePeriod = NormalizeNonNegativeMilliseconds(periodMediaMs, fallback: 1);
        safePeriod = Math.Max(1, safePeriod);
        var target = safeBase > long.MaxValue - safePeriod
            ? long.MaxValue
            : safeBase + safePeriod;
        return new(candidateId, sample, target, AudioCycleCandidateStatus.Planned);
    }

    /// <summary>返回新状态计划，不修改已有快照。</summary>
    public static AudioCycleCandidatePlan<T> WithStatus<T>(
        AudioCycleCandidatePlan<T> plan,
        AudioCycleCandidateStatus status)
    {
        ArgumentNullException.ThrowIfNull(plan);
        if (!Enum.IsDefined(status))
        {
            throw new ArgumentOutOfRangeException(nameof(status), "音频候选状态无效。");
        }

        return plan with { Status = status };
    }

    /// <summary>
    /// 计算单步调度动作：planned 只触发 prepare；prepared 在环缓待播时间覆盖目标前触发 commit；
    /// 超过目标宽限期触发 expire；committing 永不重复提交。
    /// </summary>
    public static AudioCycleCoordinatorAction GetAction<T>(
        AudioCycleCandidatePlan<T>? plan,
        double currentAbsolutePositionMs,
        double playbackRate,
        double queuedPlaybackMs = 0)
    {
        if (plan is null)
        {
            return AudioCycleCoordinatorAction.None;
        }

        var safeCurrent = NormalizeNonNegativePosition(currentAbsolutePositionMs);
        var safeRate = double.IsFinite(playbackRate) && playbackRate > 0
            ? Math.Min(playbackRate, 16d)
            : 1d;
        var safeQueued = double.IsFinite(queuedPlaybackMs) && queuedPlaybackMs > 0
            ? Math.Min(queuedPlaybackMs, double.MaxValue / safeRate)
            : 0d;
        var grace = CommitGraceMediaMs * safeRate;
        var target = (double)plan.TargetAbsolutePositionMs;

        if (safeCurrent > target + grace)
        {
            return plan.Status is AudioCycleCandidateStatus.Committing
                ? AudioCycleCoordinatorAction.None
                : AudioCycleCoordinatorAction.Expire;
        }

        if (plan.Status is AudioCycleCandidateStatus.Committing)
        {
            return AudioCycleCoordinatorAction.None;
        }

        if (plan.Status is AudioCycleCandidateStatus.Planned)
        {
            return AudioCycleCoordinatorAction.Prepare;
        }

        var mediaLead = target - safeCurrent;
        var commitLead = safeQueued * safeRate;
        return plan.Status is AudioCycleCandidateStatus.Prepared
            && mediaLead <= commitLead
            ? AudioCycleCoordinatorAction.Commit
            : AudioCycleCoordinatorAction.None;
    }

    private static long NormalizeNonNegativeMilliseconds(double value, long fallback)
    {
        if (!double.IsFinite(value) || value <= 0)
        {
            return fallback;
        }

        return value >= long.MaxValue
            ? long.MaxValue
            : Math.Max(1, checked((long)Math.Round(value, MidpointRounding.AwayFromZero)));
    }

    private static double NormalizeNonNegativePosition(double value) =>
        double.IsFinite(value) && value > 0 ? value : 0;
}
